src/include/mpir_handlemem.h

```diff
@@ -393,6 +394,58 @@ static inline void MPIR_Handle_obj_free_unsafe(MPIR_Object_alloc_t * objmem, voi
 
     obj->next = objmem->avail;
     objmem->avail = obj;
+
+    if (enable_gc) {
+        /* This option *significantly* slows down this routine.  This active
+         * garbage collection is for MPI_Info, which cannot free its blocks on
+         * finalization. */
+
+        /* Check if the objmem's linked list contains all the indirect memory
+         * segments. */
+        int num_indirect = 0;
+        {
+            /* NOTE: direct could be NULL (and directSize == 0) */
+            char *direct = (char *) objmem->direct;
+            char *directEnd = (char *) direct + objmem->direct_size * objmem->size - 1;
+
+            MPIR_Handle_common *ptr = objmem->avail;
+            while (ptr) {
+                /* If it's not direct, it is indirect.  Note that built-in
+                 * objects may not be freed. */
+                if (!(direct && (char *) ptr >= direct && (char *) ptr < directEnd)) {
+                    num_indirect++;
+                }
+                ptr = ptr->next;
+            }
+        }
+        if (num_indirect == objmem->indirect_size * HANDLE_NUM_INDICES) {
+            /* All indirect memory has been released.  Let's free all. */
+
+            /* First, avail should not contain indirect objmem.  Remove all. */
+            char *direct = (char *) objmem->direct;
+            char *directEnd = (char *) direct + objmem->direct_size * objmem->size - 1;
+
+            MPIR_Handle_common *ptr = objmem->avail;
+            MPIR_Handle_common **p_newptr = &objmem->avail;
+            while (ptr) {
+                if (direct && (char *) ptr >= direct && (char *) ptr < directEnd) {
+                    /* This ptr is direct.  Add it to p_newptr */
+                    *p_newptr = ptr;
+                    p_newptr = (MPIR_Handle_common **) & ptr->next;
+                }
+                ptr = ptr->next;
+            }
+            *p_newptr = NULL;
+
+            /* Free the real memory that is dynamically allocated. */
+            (void) MPIR_Handle_free(objmem->indirect, objmem->indirect_size);
+            MPL_VG_DESTROY_MEMPOOL(objmem);
+            /* Since freed indirect etc will be reused, it must be initialized
+             * here. */
+            objmem->indirect_size = 0;
+            objmem->indirect = NULL;
+        }
+    }
```

Comment:
I did not modify the existing mechanism since it affects the overheads of request allocation even if the cost is tiny. Do we really need to optimize MPI_Info allocation/deallocation?

Reply:
The overhead of a single increment/decrement is insignificant to worry about, right? I am more concerned about the amount of code we need to maintain. If we can simplify the code, especially when we are adding new code, I am tempted to push for that.

Reply:
If it's okay to add such a cost, there's no problem. I will add two counters for direct and indirect.

Reply:
Yeah, the cost is measured against a few branches, a lock, and at least hundreds of instructions.

Reply:
Do you also want to remove the `enable_gc` parameter, which is newly added by this PR?

If it is okay to add a branch, I can remove this parameter and make things cleaner.

Reply:
How much does a branch cost?

Reply:
I cannot say anything certain without measurement, but I guess 5 cycles or so?

Reply:
> I cannot say anything certain without measurement, but I guess 5 cycles or so?

And it is difficult to measure. Then we need to look at how much code simplification you can achieve with the branch. If we can't decide, let's first ensure we don't degrade the performance.

Reply:
So you are saying that I should prepare two implementations so that we can examine the difference, right? I will create two branches so that you can check them.

Reply:
No, that is not what I am saying. I am saying when decisions are not easy, we should lean on the conservative side. The conservative side being ensures performance first. 

In adding a single clock counter to reduce dozens of line of code with rather complex logic, that is an easy decision. Now adding a branch, I am not so sure.

Reply:
Okay, then I will keep the `enable_gc` parameter. Thanks.

Reply:
I added counters `num_freed` and `num_allocated`, which simplifies the logic in `MPIR_Handle_obj_free_unsafe()`. I did add a new branch.